homeassistant/components/apple_tv/config_flow.py

```diff
@@ -0,0 +1,406 @@
+        credentials.get(Protocol.MRP.value) is not None
+        or credentials.get(Protocol.DMAP.value) is not None
+    )
+
+
+class AppleTVConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Apple TV."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(config_entry):
+        """Get options flow for this handler."""
+        return AppleTVOptionsFlow(config_entry)
+
+    def __init__(self):
+        """Initialize a new AppleTVConfigFlow."""
+        self.scan_result = None
+        self.atv = None
+        self.protocol = None
+        self.pairing = None
+        self.credentials = {}  # Protocol -> credentials
+
+    async def async_step_reauth(self, info):
+        """Handle initial step when updating invalid credentials."""
+        await self.async_set_unique_id(info[CONF_IDENTIFIER])
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["title_placeholders"] = {"name": info[CONF_NAME]}
+
+        # pylint: disable=no-member # https://github.com/PyCQA/pylint/issues/3167
+        self.context["identifier"] = self.unique_id
+        return await self.async_step_reconfigure()
+
+    async def async_step_reconfigure(self, user_input=None):
+        """Inform user that reconfiguration is about to start."""
+        if user_input is not None:
+            return await self.async_find_device_wrapper(
+                self.async_begin_pairing, allow_exist=True
+            )
+
+        return self.async_show_form(step_id="reconfigure")
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        # Be helpful to the user and look for devices
+        if self.scan_result is None:
+            self.scan_result, _ = await device_scan(None, self.hass.loop)
+
+        errors = {}
+        default_suggestion = self._prefill_identifier()
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[DEVICE_INPUT])
+            try:
+                await self.async_find_device()
+            except DeviceNotFound:
+                errors["base"] = "device_not_found"
+            except DeviceAlreadyConfigured:
+                errors["base"] = "device_already_configured"
+            except exceptions.NoServiceError:
+                errors["base"] = "no_usable_service"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return await self.async_step_confirm()
+
+            # Use whatever the user entered as default value
+            default_suggestion = self.unique_id
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {vol.Required(DEVICE_INPUT, default=default_suggestion): str}
```

Comment:
I'd suggest providing a list of discovered devices to choose from together with a manual option. See eg the hue integration.

https://github.com/home-assistant/core/blob/343e5d64b82b61a96051b3bf8994037341489b63/homeassistant/components/hue/config_flow.py#L100-L112

Reply:
Could that be considered for a future PR instead (close in time)? I will never finish this one if new things pops up ad infinitum. I mean, current solution still works.

Reply:
It seems we already have the discovered items available so the change would basically just be changing how we generate the form. But it can be a separate PR, yes.

Reply:
It's probably not a big change. I'm just getting a bit tired with this PR, would mentally feel better to handle it separately.